src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,311 @@
+        private const int StackAllocThreshold = 128;
+
+        private readonly PipeReader _pipeReader;
+        private readonly MultipartBoundary _boundary;
+        private MultipartSectionPipeReader _currentSection;
+
+        private static ReadOnlySpan<byte> ColonDelimiter => new byte[] { (byte)':' };
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+
+        public MultipartPipeReader(string boundary, PipeReader pipeReader)
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+
+            // This stream will drain any preamble data and remove the first boundary marker. 
+            // TODO: HeadersLengthLimit can't be modified until after the constructor. 
+            _currentSection = new MultipartSectionPipeReader(_pipeReader, _boundary);
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional limit for the total response body length.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+        public async Task<MultipartSection> ReadNextSectionAsync(CancellationToken cancellationToken = default)
+        {
+            await _currentSection.DrainAsync(cancellationToken);
+
+            var headersAccumulator = new KeyValueAccumulator();
+            _boundary.ExpectLeadingCrlf = true;
+            long headersLength = 0;
+            while (true)
+            {
+                var readResult = await _pipeReader.ReadAsync(cancellationToken);
+
+                if (readResult.IsCanceled)
+                {
+                    throw new OperationCanceledException("Read was canceled");
+                }
+
+                var buffer = readResult.Buffer;
+
+                if (buffer.IsEmpty)
+                {
+                    if (readResult.IsCompleted)
+                    {
+                        return null;
+                    }
+                    // This is a buggy PipeReader implementation that returns 0 byte reads even though the PipeReader
+                    // isn't completed or canceled - still think it should continue
+                    continue;
+                }
+
+                var finishedParsing = TryParseHeadersToEnd(ref buffer, ref headersAccumulator, ref headersLength);
+                if (headersLength > DefaultHeadersLengthLimit)
+                {
+                    _pipeReader.AdvanceTo(buffer.Start, buffer.End); // free memory 
```

Comment:
This doesn't free the memory unless you decide to consume the data which this isn't doing. Change it to _pipeReader.AdvanceTo(buffer.End)

Reply:
I think it's right. Buffer gets sliced so it only represents unconsumed data.